# Day 7: The Treachery of Whales

## Part one

A giant whale has decided your submarine is its next meal, and it's much faster than you are. There's nowhere to run!

Suddenly, a swarm of crabs (each in its own tiny submarine - it's too deep for them otherwise) zooms in to rescue you! They seem to be preparing to blast a hole in the ocean floor; sensors indicate a **massive underground cave system** just beyond where they're aiming!

The crab submarines all need to be aligned before they'll have enough power to blast a large enough hole for your submarine to get through. However, it doesn't look like they'll be aligned before the whale catches you! Maybe you can help?

There's one major catch - crab submarines can only move horizontally.

You quickly make a list of **the horizontal position of each crab** (your puzzle input). Crab submarines have limited fuel, so you need to find a way to make all of their horizontal positions match while requiring them to spend as little fuel as possible.

For example, consider the following horizontal positions:

`16,1,2,0,4,2,7,1,2,14`

This means there's a crab with horizontal position 16, a crab with horizontal position 1, and so on.

Each change of 1 step in horizontal position of a single crab costs 1 fuel. You could choose any horizontal position to align them all on, but the one that costs the least fuel is horizontal position 2:

- Move from `16` to `2`: `14` fuel
- Move from `1` to `2`: `1` fuel
- Move from `2` to `2`: `0` fuel
- Move from `0` to `2`: `2` fuel
- Move from `4` to `2`: `2` fuel
- Move from `2` to `2`: `0` fuel
- Move from `7` to `2`: `5` fuel
- Move from `1` to `2`: `1` fuel
- Move from `2` to `2`: `0` fuel
- Move from `14` to `2`: `12` fuel

This costs a total of **`37`** fuel. This is the cheapest possible outcome; more expensive outcomes include aligning at position `1` (`41` fuel), position `3` (`39` fuel), or position `10` (`71` fuel).

Determine the horizontal position that the crabs can align to using the least fuel possible. **How much fuel must they spend to align to that position?**

In [62]:
crab_positions = []
example_crab_positions = [16, 1, 2, 0, 4, 2, 7, 1, 2, 14]

with open('data/input.txt', 'r') as f:
    line = f.read().splitlines()[0]
    for crab_position in line.split(','):
        crab_positions.append(int(crab_position))

crab_positions[:10]

[1101, 1, 29, 67, 1102, 0, 1, 65, 1008, 65]

In [63]:
import statistics

statistics.median(crab_positions)

print("median low: %s - median high: %s" % (statistics.median_low(crab_positions), statistics.median_high(crab_positions) ))

median low: 328 - median high: 328


In [64]:
def total_fuel_cost(crab_positions, position):
    sum_fuel_costs = 0
    for crab_position in crab_positions:
        if crab_position > position:
            sum_fuel_costs += (crab_position - position)
        else:
            sum_fuel_costs += (position - crab_position)
            
    return sum_fuel_costs

total_fuel_cost(crab_positions, 328)

339321

Your puzzle answer was `339321`.

## Part two

The crabs don't seem interested in your proposed solution. Perhaps you misunderstand crab engineering?

As it turns out, crab submarine engines don't burn fuel at a constant rate. Instead, each change of `1` step in horizontal position costs `1` more unit of fuel than the last: the first step costs `1`, the second step costs `2`, the third step costs `3`, and so on.

As each crab moves, moving further becomes more expensive. This changes the best horizontal position to align them all on; in the example above, this becomes `5`:

- Move from `16` to `5`: `66` fuel
- Move from `1` to `5`: `10` fuel
- Move from `2` to `5`: `6` fuel
- Move from `0` to `5`: `15` fuel
- Move from `4` to `5`: `1` fuel
- Move from `2` to `5`: `6` fuel
- Move from `7` to `5`: `3` fuel
- Move from `1` to `5`: `10` fuel
- Move from `2` to `5`: `6` fuel
- Move from `14` to `5`: `45` fuel

This costs a total of `168` fuel. This is the new cheapest possible outcome; the old alignment position (`2`) now costs `206` fuel instead.

Determine the horizontal position that the crabs can align to using the least fuel possible so they can make you an escape route! **How much fuel must they spend to align to that position?**

In [65]:
import statistics

statistics.mean(crab_positions)

471.613

In [66]:
def total_fuel_cost_ex2(crab_positions, position):
    sum_fuel_costs = 0
    for crab_position in crab_positions:
        sum_fuel_costs += fuel_cost_ex2(crab_position, position)
            
    return sum_fuel_costs

def fuel_cost_ex2(crab_position, position):
    sum_fuel_cost = 0
    distance = 0
    if crab_position > position:
        distance = crab_position - position
    else:
        distance = position - crab_position
        
    for i in range(0, distance + 1):
        sum_fuel_cost += i
            
    return sum_fuel_cost

total_fuel_cost_ex2(example_crab_positions, 5)

168

In [67]:
print("fuel cost 470: %s" % (total_fuel_cost_ex2(crab_positions, 470), ))
print("fuel cost 471: %s" % (total_fuel_cost_ex2(crab_positions, 471), ))
print("fuel cost 472: %s" % (total_fuel_cost_ex2(crab_positions, 472), ))

total_fuel_cost_ex2(crab_positions, 471)

fuel cost 470: 95477241
fuel cost 471: 95476244
fuel cost 472: 95476248


95476244

Your puzzle answer was `95476244`.